File: 1130-last-stone-weight-ii/1130-last-stone-weight-ii.py
class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        stonesSum = sum(stones)
        target = stonesSum // 2
        dp = [[0 for _ in range(target + 1)] for _ in range(len(stones))]

        for i in range(len(stones) - 1, -1, -1):
            for total in range(target, -1, -1):
                if i + 1 >= len(stones):
                    dp[i][total] = max(
                        total,
                        total + stones[i] if total + stones[i] <= target else -inf
                    )
                else:
                    dp[i][total] = max(
                        dp[i + 1][total],
                        dp[i + 1][total + stones[i]] if total + stones[i] <= target else -inf
                    )


        S = dp[0][0]

        return stonesSum - 2 * S

File: 1130-last-stone-weight-ii/1130-last-stone-weight-ii.cpp
class Solution {
public:
    int lastStoneWeightII(vector<int>& stones) {
        int sum = accumulate(stones.begin(), stones.end(), 0);
        int target = sum / 2;

        vector<vector<long>> memo(stones.size() + 1, vector<long>(sum + 1, -1));

        long ans = dp(stones, 0, 0, target, memo);

        return sum - 2 * ans;
    }

    int dp(vector<int>& stones, int i, long current, int target, vector<vector<long>>& memo) {
        if (memo[i][current] != -1) return memo[i][current];

        if (i >= stones.size()) return current;

        long without = dp(stones, i + 1, current, target, memo);
        long with = current + stones[i] <= target ? dp(stones, i + 1, current + stones[i], target, memo) : 0;

        long ans = max(without, with);
        memo[i][current] = ans;

        return ans;
    }
};